### ISSUE TITLE
PubSub: freed memory is accessed during writerGroup removal

### ISSUE THREAD

**Issue by MichalTyndel** [open62541/open62541#6261]
## Description

Valgrind (memcheck) detects accessing freed memory when removing writerGroups. Looks like the delayed callback, that is supposed to remove the writerGroup, is called when the structure is already deleted.  
Problem occurs when the PubSub elements are removed "bottom-up" i.e. dsw, then wg, finally connection.

## Background Information / Reproduction Steps

[example1.txt](https://github.com/open62541/open62541/files/14124627/example1.txt)
Attached is example code (c file stored as txt), that can be used to reproduce the issue. It's a mock application that establishes pubSub communication locally, then removes all PubSub objects:
- Create DataSetWriter -> DataSetReader pair that exchanges a single variable using UDP over localhost  
- Disable communication
- Remove all pubSub objects in order DSR, RG, DSW, WG, CONN, PDS.

Memcheck detects multiple invalid reads when running this example program:
![obraz](https://github.com/open62541/open62541/assets/99327189/596860dc-f4ca-4840-b76f-50815dc7ade0)

Full Memcheck log is attached.
[wgRemoval.log](https://github.com/open62541/open62541/files/14124628/wgRemoval.log)

## Checklist

 - [x] open62541 Version (release number or git tag): v1.4.0-rc2-162-gcc806cb47
 - [x] Self-contained code example attached

**Comment by JohannesKauffmann** [comment]
I've checked out open62541 at `cc806cb47`, built the library with `-DUA_ENABLE_AMALGAMATION=ON`, changed `#include open62541/server.h` to just `#include open62541.h`, but can't compile your example (also without amalgamation):
```
$ gcc -O2 -Wall -Wextra -fsanitize=address example1.c -I /usr/local/include/  -L /usr/local/lib/ -lopen62541.a
example1.c: In function ‘addInt32Variable’:
example1.c:28:53: error: ‘UA_ACCESSLEVELTYPE_CURRENTREAD’ undeclared (first use in this function)
   28 |                                      .accessLevel = UA_ACCESSLEVELTYPE_CURRENTREAD | UA_ACCESSLEVELTYPE_CURRENTWRITE};
      |                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
example1.c:28:53: note: each undeclared identifier is reported only once for each function it appears in
example1.c:28:86: error: ‘UA_ACCESSLEVELTYPE_CURRENTWRITE’ undeclared (first use in this function)
   28 |                                      .accessLevel = UA_ACCESSLEVELTYPE_CURRENTREAD | UA_ACCESSLEVELTYPE_CURRENTWRITE};
      |                                                                                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
example1.c: In function ‘addConnection’:
example1.c:37:5: error: unknown type name ‘UA_PubSubConnectionConfig’; did you mean ‘UA_ConnectionConfig’?
   37 |     UA_PubSubConnectionConfig cfg = {0};
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~
      |     UA_ConnectionConfig
example1.c:39:5: error: unknown type name ‘UA_NetworkAddressUrlDataType’
   39 |     UA_NetworkAddressUrlDataType addrVal = {.url = UA_STRING((char *)((uintptr_t)addr))};
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
example1.c:39:45: error: field name not in record or union initializer
   39 |     UA_NetworkAddressUrlDataType addrVal = {.url = UA_STRING((char *)((uintptr_t)addr))};
      |                                             ^
example1.c:39:45: note: (near initialization for ‘addrVal’)
example1.c:39:52: error: incompatible types when initializing type ‘int’ using type ‘UA_String’
   39 |     UA_NetworkAddressUrlDataType addrVal = {.url = UA_STRING((char *)((uintptr_t)addr))};
      |                                                    ^~~~~~~~~
example1.c:40:30: error: request for member ‘address’ in something not a structure or union
   40 |     UA_Variant_setScalar(&cfg.address, &addrVal, &UA_TYPES[UA_TYPES_NETWORKADDRESSURLDATATYPE]);
      |                              ^
example1.c:40:60: error: ‘UA_TYPES_NETWORKADDRESSURLDATATYPE’ undeclared (first use in this function); did you mean ‘UA_NS0ID_NETWORKADDRESSURLDATATYPE’?
   40 |     UA_Variant_setScalar(&cfg.address, &addrVal, &UA_TYPES[UA_TYPES_NETWORKADDRESSURLDATATYPE]);
      |                                                            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                                            UA_NS0ID_NETWORKADDRESSURLDATATYPE
example1.c:42:8: error: request for member ‘name’ in something not a structure or union
   42 |     cfg.name = UA_STRING((char *)((uintptr_t)name));
      |        ^
example1.c:43:8: error: request for member ‘publisherId’ in something not a structure or union
   43 |     cfg.publisherId.uint64 = pubId;
      |        ^
example1.c:44:8: error: request for member ‘publisherIdType’ in something not a structure or union
   44 |     cfg.publisherIdType = UA_PUBLISHERIDTYPE_UINT64;
      |        ^
example1.c:44:27: error: ‘UA_PUBLISHERIDTYPE_UINT64’ undeclared (first use in this function)
   44 |     cfg.publisherIdType = UA_PUBLISHERIDTYPE_UINT64;
      |                           ^~~~~~~~~~~~~~~~~~~~~~~~~
example1.c:45:8: error: request for member ‘transportProfileUri’ in something not a structure or union
   45 |     cfg.transportProfileUri = UA_STRING("http://opcfoundation.org/UA-Profile/Transport/pubsub-udp-uadp");
      |        ^
example1.c:48:25: warning: implicit declaration of function ‘UA_Server_addPubSubConnection’; did you mean ‘UA_Server_addReverseConnect’? [-Wimplicit-function-declaration]
   48 |     UA_StatusCode sts = UA_Server_addPubSubConnection(srv, &cfg, &ret);
      |                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                         UA_Server_addReverseConnect
example1.c: In function ‘addWriterGroup’:
example1.c:55:5: error: unknown type name ‘UA_WriterGroupConfig’
   55 |     UA_WriterGroupConfig cfg = {0};
      |     ^~~~~~~~~~~~~~~~~~~~
example1.c:56:8: error: request for member ‘name’ in something not a structure or union
   56 |     cfg.name = UA_STRING((char *)((uintptr_t)name));
      |        ^
example1.c:57:8: error: request for member ‘writerGroupId’ in something not a structure or union
   57 |     cfg.writerGroupId = wgId;
      |        ^
example1.c:58:8: error: request for member ‘enabled’ in something not a structure or union
   58 |     cfg.enabled = true;
      |        ^
example1.c:59:8: error: request for member ‘keepAliveTime’ in something not a structure or union
   59 |     cfg.keepAliveTime = 10;
      |        ^
example1.c:60:8: error: request for member ‘publishingInterval’ in something not a structure or union
   60 |     cfg.publishingInterval = 10;
      |        ^
example1.c:61:8: error: request for member ‘securityMode’ in something not a structure or union
   61 |     cfg.securityMode = UA_MESSAGESECURITYMODE_NONE;
      |        ^
example1.c:63:5: error: unknown type name ‘UA_UadpWriterGroupMessageDataType’
   63 |     UA_UadpWriterGroupMessageDataType ms = {.dataSetOrdering = UA_DATASETORDERINGTYPE_UNDEFINED,
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
example1.c:63:45: error: field name not in record or union initializer
   63 |     UA_UadpWriterGroupMessageDataType ms = {.dataSetOrdering = UA_DATASETORDERINGTYPE_UNDEFINED,
      |                                             ^
example1.c:63:45: note: (near initialization for ‘ms’)
example1.c:63:64: error: ‘UA_DATASETORDERINGTYPE_UNDEFINED’ undeclared (first use in this function); did you mean ‘UA_NS0ID_DATASETORDERINGTYPE_ENUMSTRINGS’?
   63 |     UA_UadpWriterGroupMessageDataType ms = {.dataSetOrdering = UA_DATASETORDERINGTYPE_UNDEFINED,
      |                                                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                                                UA_NS0ID_DATASETORDERINGTYPE_ENUMSTRINGS
example1.c:64:45: error: field name not in record or union initializer
   64 |                                             .networkMessageContentMask = 0x41,
      |                                             ^
example1.c:64:45: note: (near initialization for ‘ms’)
example1.c:64:74: warning: excess elements in scalar initializer
   64 |                                             .networkMessageContentMask = 0x41,
      |                                                                          ^~~~
example1.c:64:74: note: (near initialization for ‘ms’)
example1.c:65:45: error: field name not in record or union initializer
   65 |                                             .groupVersion = 1};
      |                                             ^
example1.c:65:45: note: (near initialization for ‘ms’)
example1.c:65:61: warning: excess elements in scalar initializer
   65 |                                             .groupVersion = 1};
      |                                                             ^
example1.c:65:61: note: (near initialization for ‘ms’)
example1.c:66:37: error: request for member ‘messageSettings’ in something not a structure or union
   66 |     UA_ExtensionObject_setValue(&cfg.messageSettings, &ms, &UA_TYPES[UA_TYPES_UADPWRITERGROUPMESSAGEDATATYPE]);
      |                                     ^
example1.c:66:70: error: ‘UA_TYPES_UADPWRITERGROUPMESSAGEDATATYPE’ undeclared (first use in this function); did you mean ‘UA_NS0ID_UADPWRITERGROUPMESSAGEDATATYPE’?
   66 |     UA_ExtensionObject_setValue(&cfg.messageSettings, &ms, &UA_TYPES[UA_TYPES_UADPWRITERGROUPMESSAGEDATATYPE]);
      |                                                                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                                                      UA_NS0ID_UADPWRITERGROUPMESSAGEDATATYPE
example1.c:68:5: error: unknown type name ‘UA_DatagramWriterGroupTransport2DataType’
   68 |     UA_DatagramWriterGroupTransport2DataType ts = {};
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
example1.c:69:5: error: unknown type name ‘UA_NetworkAddressUrlDataType’
   69 |     UA_NetworkAddressUrlDataType addrVal = {.url = UA_STRING((char *)((uintptr_t)addr))};
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
example1.c:69:45: error: field name not in record or union initializer
   69 |     UA_NetworkAddressUrlDataType addrVal = {.url = UA_STRING((char *)((uintptr_t)addr))};
      |                                             ^
example1.c:69:45: note: (near initialization for ‘addrVal’)
example1.c:69:52: error: incompatible types when initializing type ‘int’ using type ‘UA_String’
   69 |     UA_NetworkAddressUrlDataType addrVal = {.url = UA_STRING((char *)((uintptr_t)addr))};
      |                                                    ^~~~~~~~~
example1.c:70:36: error: request for member ‘address’ in something not a structure or union
   70 |     UA_ExtensionObject_setValue(&ts.address, &addrVal, &UA_TYPES[UA_TYPES_NETWORKADDRESSURLDATATYPE]);
      |                                    ^
example1.c:70:66: error: ‘UA_TYPES_NETWORKADDRESSURLDATATYPE’ undeclared (first use in this function); did you mean ‘UA_NS0ID_NETWORKADDRESSURLDATATYPE’?
   70 |     UA_ExtensionObject_setValue(&ts.address, &addrVal, &UA_TYPES[UA_TYPES_NETWORKADDRESSURLDATATYPE]);
      |                                                                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                                                  UA_NS0ID_NETWORKADDRESSURLDATATYPE
example1.c:71:37: error: request for member ‘transportSettings’ in something not a structure or union
   71 |     UA_ExtensionObject_setValue(&cfg.transportSettings, &ts, &UA_TYPES[UA_TYPES_DATAGRAMWRITERGROUPTRANSPORT2DATATYPE]);
      |                                     ^
example1.c:71:72: error: ‘UA_TYPES_DATAGRAMWRITERGROUPTRANSPORT2DATATYPE’ undeclared (first use in this function); did you mean ‘UA_NS0ID_DATAGRAMWRITERGROUPTRANSPORT2DATATYPE’?
   71 |     UA_ExtensionObject_setValue(&cfg.transportSettings, &ts, &UA_TYPES[UA_TYPES_DATAGRAMWRITERGROUPTRANSPORT2DATATYPE]);
      |                                                                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                                                        UA_NS0ID_DATAGRAMWRITERGROUPTRANSPORT2DATATYPE
example1.c:73:8: error: request for member ‘maxEncapsulatedDataSetMessageCount’ in something not a structure or union
   73 |     cfg.maxEncapsulatedDataSetMessageCount = 255;
      |        ^
example1.c:74:8: error: request for member ‘encodingMimeType’ in something not a structure or union
   74 |     cfg.encodingMimeType = UA_PUBSUB_ENCODING_UADP;
      |        ^
example1.c:74:28: error: ‘UA_PUBSUB_ENCODING_UADP’ undeclared (first use in this function)
   74 |     cfg.encodingMimeType = UA_PUBSUB_ENCODING_UADP;
      |                            ^~~~~~~~~~~~~~~~~~~~~~~
example1.c:77:25: warning: implicit declaration of function ‘UA_Server_addWriterGroup’; did you mean ‘UA_Server_addViewNode’? [-Wimplicit-function-declaration]
   77 |     UA_StatusCode sts = UA_Server_addWriterGroup(srv, connId, &cfg, &ret);
      |                         ^~~~~~~~~~~~~~~~~~~~~~~~
      |                         UA_Server_addViewNode
example1.c: In function ‘addPds’:
example1.c:85:5: error: unknown type name ‘UA_PublishedDataSetConfig’
   85 |     UA_PublishedDataSetConfig cfg = {0};
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~
example1.c:86:8: error: request for member ‘name’ in something not a structure or union
   86 |     cfg.name = UA_STRING((char *)((uintptr_t)name));
      |        ^
example1.c:87:8: error: request for member ‘publishedDataSetType’ in something not a structure or union
   87 |     cfg.publishedDataSetType = UA_PUBSUB_DATASET_PUBLISHEDITEMS;
      |        ^
example1.c:87:32: error: ‘UA_PUBSUB_DATASET_PUBLISHEDITEMS’ undeclared (first use in this function)
   87 |     cfg.publishedDataSetType = UA_PUBSUB_DATASET_PUBLISHEDITEMS;
      |                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
example1.c:90:5: error: unknown type name ‘UA_AddPublishedDataSetResult’
   90 |     UA_AddPublishedDataSetResult result = UA_Server_addPublishedDataSet(srv, &cfg, &ret);
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
example1.c:90:43: warning: implicit declaration of function ‘UA_Server_addPublishedDataSet’ [-Wimplicit-function-declaration]
   90 |     UA_AddPublishedDataSetResult result = UA_Server_addPublishedDataSet(srv, &cfg, &ret);
      |                                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/local/include/open62541.h:328,
                 from example1.c:2:
example1.c:91:39: error: request for member ‘addResult’ in something not a structure or union
   91 |     assert(UA_StatusCode_isGood(result.addResult));
      |                                       ^
example1.c:91:39: error: request for member ‘addResult’ in something not a structure or union
   91 |     assert(UA_StatusCode_isGood(result.addResult));
      |                                       ^
example1.c:94:5: error: unknown type name ‘UA_DataSetFieldConfig’
   94 |     UA_DataSetFieldConfig fieldConfig = {.dataSetFieldType = UA_PUBSUB_DATASETFIELD_VARIABLE};
      |     ^~~~~~~~~~~~~~~~~~~~~
example1.c:94:42: error: field name not in record or union initializer
   94 |     UA_DataSetFieldConfig fieldConfig = {.dataSetFieldType = UA_PUBSUB_DATASETFIELD_VARIABLE};
      |                                          ^
example1.c:94:42: note: (near initialization for ‘fieldConfig’)
example1.c:94:62: error: ‘UA_PUBSUB_DATASETFIELD_VARIABLE’ undeclared (first use in this function)
   94 |     UA_DataSetFieldConfig fieldConfig = {.dataSetFieldType = UA_PUBSUB_DATASETFIELD_VARIABLE};
      |                                                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
example1.c:95:5: error: unknown type name ‘UA_PublishedVariableDataType’
   95 |     UA_PublishedVariableDataType *varAttr = &fieldConfig.field.variable.publishParameters;
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
example1.c:95:57: error: request for member ‘field’ in something not a structure or union
   95 |     UA_PublishedVariableDataType *varAttr = &fieldConfig.field.variable.publishParameters;
      |                                                         ^
example1.c:96:12: error: request for member ‘attributeId’ in something not a structure or union
   96 |     varAttr->attributeId = UA_ATTRIBUTEID_VALUE;
      |            ^~
example1.c:97:12: error: request for member ‘publishedVariable’ in something not a structure or union
   97 |     varAttr->publishedVariable = var;
      |            ^~
example1.c:99:5: error: unknown type name ‘UA_DataSetFieldResult’
   99 |     UA_DataSetFieldResult addFieldResult = UA_Server_addDataSetField(srv, ret, &fieldConfig, NULL);
      |     ^~~~~~~~~~~~~~~~~~~~~
example1.c:99:44: warning: implicit declaration of function ‘UA_Server_addDataSetField’; did you mean ‘UA_Server_addDataTypeNode’? [-Wimplicit-function-declaration]
   99 |     UA_DataSetFieldResult addFieldResult = UA_Server_addDataSetField(srv, ret, &fieldConfig, NULL);
      |                                            ^~~~~~~~~~~~~~~~~~~~~~~~~
      |                                            UA_Server_addDataTypeNode
example1.c:100:47: error: request for member ‘result’ in something not a structure or union
  100 |     assert(UA_StatusCode_isGood(addFieldResult.result));
      |                                               ^
example1.c:100:47: error: request for member ‘result’ in something not a structure or union
  100 |     assert(UA_StatusCode_isGood(addFieldResult.result));
      |                                               ^
example1.c:99:27: warning: variable ‘addFieldResult’ set but not used [-Wunused-but-set-variable]
   99 |     UA_DataSetFieldResult addFieldResult = UA_Server_addDataSetField(srv, ret, &fieldConfig, NULL);
      |                           ^~~~~~~~~~~~~~
example1.c:90:34: warning: variable ‘result’ set but not used [-Wunused-but-set-variable]
   90 |     UA_AddPublishedDataSetResult result = UA_Server_addPublishedDataSet(srv, &cfg, &ret);
      |                                  ^~~~~~
example1.c: In function ‘addWriter’:
example1.c:107:5: error: unknown type name ‘UA_DataSetWriterConfig’; did you mean ‘UA_ServerConfig’?
  107 |     UA_DataSetWriterConfig cfg = {0};
      |     ^~~~~~~~~~~~~~~~~~~~~~
      |     UA_ServerConfig
example1.c:108:8: error: request for member ‘name’ in something not a structure or union
  108 |     cfg.name = UA_STRING((char *)((uintptr_t)name));
      |        ^
example1.c:109:8: error: request for member ‘dataSetWriterId’ in something not a structure or union
  109 |     cfg.dataSetWriterId = dswId;
      |        ^
example1.c:110:8: error: request for member ‘keyFrameCount’ in something not a structure or union
  110 |     cfg.keyFrameCount = 1;
      |        ^
example1.c:111:8: error: request for member ‘dataSetName’ in something not a structure or union
  111 |     cfg.dataSetName = UA_STRING((char *)((uintptr_t)pdsName));
      |        ^
example1.c:112:8: error: request for member ‘dataSetFieldContentMask’ in something not a structure or union
  112 |     cfg.dataSetFieldContentMask = 0;
      |        ^
example1.c:114:5: error: unknown type name ‘UA_UadpDataSetWriterMessageDataType’
  114 |     UA_UadpDataSetWriterMessageDataType ms = {.dataSetMessageContentMask = 0x35, .networkMessageNumber = 1};
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
example1.c:114:47: error: field name not in record or union initializer
  114 |     UA_UadpDataSetWriterMessageDataType ms = {.dataSetMessageContentMask = 0x35, .networkMessageNumber = 1};
      |                                               ^
example1.c:114:47: note: (near initialization for ‘ms’)
example1.c:114:82: error: field name not in record or union initializer
  114 |     UA_UadpDataSetWriterMessageDataType ms = {.dataSetMessageContentMask = 0x35, .networkMessageNumber = 1};
      |                                                                                  ^
example1.c:114:82: note: (near initialization for ‘ms’)
example1.c:114:106: warning: excess elements in scalar initializer
  114 |     UA_UadpDataSetWriterMessageDataType ms = {.dataSetMessageContentMask = 0x35, .networkMessageNumber = 1};
      |                                                                                                          ^
example1.c:114:106: note: (near initialization for ‘ms’)
example1.c:115:37: error: request for member ‘messageSettings’ in something not a structure or union
  115 |     UA_ExtensionObject_setValue(&cfg.messageSettings, &ms, &UA_TYPES[UA_TYPES_UADPDATASETWRITERMESSAGEDATATYPE]);
      |                                     ^
example1.c:115:70: error: ‘UA_TYPES_UADPDATASETWRITERMESSAGEDATATYPE’ undeclared (first use in this function); did you mean ‘UA_NS0ID_UADPDATASETWRITERMESSAGEDATATYPE’?
  115 |     UA_ExtensionObject_setValue(&cfg.messageSettings, &ms, &UA_TYPES[UA_TYPES_UADPDATASETWRITERMESSAGEDATATYPE]);
      |                                                                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                                                      UA_NS0ID_UADPDATASETWRITERMESSAGEDATATYPE
example1.c:118:25: warning: implicit declaration of function ‘UA_Server_addDataSetWriter’; did you mean ‘UA_Server_addDataTypeNode’? [-Wimplicit-function-declaration]
  118 |     UA_StatusCode sts = UA_Server_addDataSetWriter(srv, wgId, pdsNodeId, &cfg, &ret);
      |                         ^~~~~~~~~~~~~~~~~~~~~~~~~~
      |                         UA_Server_addDataTypeNode
example1.c: In function ‘addReaderGroup’:
example1.c:125:5: error: unknown type name ‘UA_ReaderGroupConfig’
  125 |     UA_ReaderGroupConfig cfg = {0};
      |     ^~~~~~~~~~~~~~~~~~~~
example1.c:126:8: error: request for member ‘name’ in something not a structure or union
  126 |     cfg.name = UA_STRING((char *)((uintptr_t)name));
      |        ^
example1.c:127:8: error: request for member ‘securityMode’ in something not a structure or union
  127 |     cfg.securityMode = UA_MESSAGESECURITYMODE_NONE;
      |        ^
example1.c:128:8: error: request for member ‘encodingMimeType’ in something not a structure or union
  128 |     cfg.encodingMimeType = UA_PUBSUB_ENCODING_UADP;
      |        ^
example1.c:128:28: error: ‘UA_PUBSUB_ENCODING_UADP’ undeclared (first use in this function)
  128 |     cfg.encodingMimeType = UA_PUBSUB_ENCODING_UADP;
      |                            ^~~~~~~~~~~~~~~~~~~~~~~
example1.c:131:25: warning: implicit declaration of function ‘UA_Server_addReaderGroup’; did you mean ‘UA_Server_addReference’? [-Wimplicit-function-declaration]
  131 |     UA_StatusCode sts = UA_Server_addReaderGroup(srv, connId, &cfg, &ret);
      |                         ^~~~~~~~~~~~~~~~~~~~~~~~
      |                         UA_Server_addReference
example1.c: At top level:
example1.c:137:59: error: unknown type name ‘UA_DataSetMetaDataType’
  137 |                            uint16_t wgId, uint16_t dswId, UA_DataSetMetaDataType metaData, UA_NodeId var)
      |                                                           ^~~~~~~~~~~~~~~~~~~~~~
example1.c: In function ‘pubSubComm’:
example1.c:184:5: warning: implicit declaration of function ‘UA_Server_enableWriterGroup’ [-Wimplicit-function-declaration]
  184 |     UA_Server_enableWriterGroup(srv, wgId);
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~
example1.c:191:5: warning: implicit declaration of function ‘UA_Server_disableDataSetWriter’ [-Wimplicit-function-declaration]
  191 |     UA_Server_disableDataSetWriter(srv, dswId);
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
example1.c:195:5: warning: implicit declaration of function ‘UA_Server_enableReaderGroup’ [-Wimplicit-function-declaration]
  195 |     UA_Server_enableReaderGroup(srv, rgId);
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~
example1.c:198:5: error: unknown type name ‘UA_DataSetMetaDataType’
  198 |     UA_DataSetMetaDataType metaData = {0};
      |     ^~~~~~~~~~~~~~~~~~~~~~
example1.c:199:5: warning: implicit declaration of function ‘UA_Server_getPublishedDataSetMetaData’ [-Wimplicit-function-declaration]
  199 |     UA_Server_getPublishedDataSetMetaData(srv, pdsId, &metaData);
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
example1.c:200:23: warning: implicit declaration of function ‘addReader’ [-Wimplicit-function-declaration]
  200 |     UA_NodeId dsrId = addReader(srv, rgId, "dsr", 1, 1, 1, metaData, inVarNodeId);
      |                       ^~~~~~~~~
example1.c:200:23: error: invalid initializer
example1.c:201:5: warning: implicit declaration of function ‘UA_Server_disableDataSetReader’ [-Wimplicit-function-declaration]
  201 |     UA_Server_disableDataSetReader(srv, dsrId);
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
example1.c:202:5: warning: implicit declaration of function ‘UA_DataSetMetaDataType_clear’; did you mean ‘UA_TimeZoneDataType_clear’? [-Wimplicit-function-declaration]
  202 |     UA_DataSetMetaDataType_clear(&metaData);
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |     UA_TimeZoneDataType_clear
example1.c:207:11: warning: implicit declaration of function ‘UA_Server_enableDataSetReader’ [-Wimplicit-function-declaration]
  207 |     sts = UA_Server_enableDataSetReader(srv, dsrId);
      |           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
example1.c:209:11: warning: implicit declaration of function ‘UA_Server_enableDataSetWriter’ [-Wimplicit-function-declaration]
  209 |     sts = UA_Server_enableDataSetWriter(srv, dswId);
      |           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
example1.c:231:5: warning: implicit declaration of function ‘UA_Server_removeDataSetReader’ [-Wimplicit-function-declaration]
  231 |     UA_Server_removeDataSetReader(srv, dsrId);
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
example1.c:232:5: warning: implicit declaration of function ‘UA_Server_removeReaderGroup’; did you mean ‘UA_Server_removeReverseConnect’? [-Wimplicit-function-declaration]
  232 |     UA_Server_removeReaderGroup(srv, rgId);
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~
      |     UA_Server_removeReverseConnect
example1.c:233:5: warning: implicit declaration of function ‘UA_Server_removeDataSetWriter’ [-Wimplicit-function-declaration]
  233 |     UA_Server_removeDataSetWriter(srv, dswId);
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
example1.c:234:5: warning: implicit declaration of function ‘UA_Server_removeWriterGroup’ [-Wimplicit-function-declaration]
  234 |     UA_Server_removeWriterGroup(srv, wgId);
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~
example1.c:235:5: warning: implicit declaration of function ‘UA_Server_removePubSubConnection’; did you mean ‘UA_Server_removeReverseConnect’? [-Wimplicit-function-declaration]
  235 |     UA_Server_removePubSubConnection(srv, connId);
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |     UA_Server_removeReverseConnect
example1.c:236:5: warning: implicit declaration of function ‘UA_Server_removePublishedDataSet’ [-Wimplicit-function-declaration]
  236 |     UA_Server_removePublishedDataSet(srv, pdsId);
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by MichalTyndel** [comment]
@JohannesKauffmann sorry for late reply. Those defines should be visible in the example. They are deifined in types_generated.h, which is included by types.h, which in turn is included by server.h.
Anyway here is CMakeLists file I used to build the example:
[CMakeLists.txt](https://github.com/open62541/open62541/files/14268340/CMakeLists.txt)

**Comment by jpfr** [comment]
The fix arrives in #6284
